File: countIslands.cpp
/*

Intuition: Consider the problem as a connected component problem. This problem can be done in both bfs and dfs manner.
To avoid visiting a node again, change its state.
//////////////////////////////
BFS Time Complexity: O(2*MxN) = O(MxN) 
BFS Space Complexity: O(min(row, cols))
//////////////////////////////
DFS Time Complexity: 
DFS Space Complexity:

*/
//BFS
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        
        int islands = 0;
        vector <pair<int,int>> queue; 
        vector <vector<int>> dirs = {{-1,0}, {0,-1}, {0,1}, {1,0}};
        
        int rows = grid.size();
        int cols = grid[0].size();
        cout <<rows << cols;
        for (int i = 0; i < rows; i++){
            for(int j = 0; j <cols; j++){
                cout << grid[i][j]<< endl;
                if (grid[i][j] == '1'){                    
                    islands += 1;
                    grid[i][j] = '0';
                    queue.push_back({i,j});
                    while (queue.size() != 0){
                    
                        pair <int, int> currentIsland = queue.front();
                        queue.erase(queue.begin());
                        int currentI = currentIsland.first;
                        int currentJ = currentIsland.second;
                        
                        for (auto dir: dirs){
                            int newI = currentI + dir[0];
                            int newJ = currentJ + dir[1];
                            if (newI >= 0 and newI < rows and newJ >=0 and newJ <cols and grid[newI][newJ] == '1' ){
                                  
                                grid[newI][newJ] = '0';
                                queue.push_back({newI,newJ});
                            }
                        }
                    
                    }
                
                }
            
            }
        
        }
        
        
        
        return islands;
        
    }
};

//DFS
class Solution {
public:
    vector <vector<int>> dirs;
    int numIslands(vector<vector<char>>& grid) {
        int islands = 0;
        dirs = {{-1,0}, {0,-1}, {0,1}, {1,0}};
        
        int rows = grid.size();
        int cols = grid[0].size();
        cout <<rows << cols;
        for (int i = 0; i < rows; i++){
            for(int j = 0; j <cols; j++){
                cout << grid[i][j]<< endl;
                if (grid[i][j] == '1'){                    
                    islands += 1;
                    grid[i][j] = '0';
                    dfs(grid, i, j,  rows,  cols);
                    
                
                }
            
            }
        
        }
        
        return islands;
        
    }
    
    void dfs(vector<vector<char>>& grid, int currI, int currJ, int rows, int cols){

        for (auto dir: dirs){
            int newI = currI + dir[0];
            int newJ = currJ + dir[1];
            if (newI >= 0 and newI < rows and newJ >=0 and newJ <cols and grid[newI][newJ] == '1' ){
                grid[newI][newJ] = '0';
                dfs(grid, newI, newJ, rows, cols);
            }
        }

    
    }
};

File: countislands.py
#Time Complexity: O(mn)
#Space Complexity: O(mn)
class Solution:
    dirns = [[1,0],[-1,0],[0,1],[0,-1]]
    def numIslands(self, grid: List[List[str]]) -> int:
        
        m = len(grid)
        n = len(grid[0])
        count = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    
                    self.bfs(i,j,grid,m,n)
                    count += 1
        return count
                    
    def bfs(self,i,j,grid,m,n):
        q = deque()
        q.append([i,j])
        
        while q:
            r,c = q.popleft()
            for dirn in self.dirns:
                nr = r + dirn[0]
                nc = c + dirn[1]
                if nr>=0 and nr<m and nc>=0 and nc<n and grid[nr][nc] == '1':
                    q.append([nr,nc])
                    grid[nr][nc] = '2'
                    
                    
                    
                
                
                